src/main/java/org/apache/datasketches/tuple/Intersection.java

```diff
@@ -19,98 +19,176 @@
     }
-    // assumes that constructor of QuickSelectSketch bumps the requested size up to the nearest power of 2
-    if (isFirstCall) {
-      sketch_ = new QuickSelectSketch<>(sketchIn.getRetainedEntries(), ResizeFactor.X1.lg(), null);
-      final SketchIterator<S> it = sketchIn.iterator();
-      while (it.next()) {
-        final S summary = (S)it.getSummary().copy();
-        sketch_.insert(it.getKey(), summary);
-      }
-    } else {
-      if (sketch_ == null) {
+    // input sketch will have valid entries > 0
+
+    if (firstCall) {
+      final Sketch<S> firstSketch = sketchIn;
+      //Copy firstSketch data into local instance hashTables_
+      hashTables_.fromSketch(firstSketch);
+    }
+
+    //Next Call
+    else {
+      if (hashTables_.count_ == 0) {
         return;
       }
-      final int matchSize = min(sketch_.getRetainedEntries(), sketchIn.getRetainedEntries());
-      final long[] matchKeys = new long[matchSize];
+      final Sketch<S> nextSketch = sketchIn;
+      //Match nextSketch data with local instance data, filtering by theta
+      final int maxMatchSize = min(hashTables_.count_, nextSketch.getRetainedEntries());
+
+      final long[] matchKeys = new long[maxMatchSize];
       S[] matchSummaries = null;
       int matchCount = 0;
-      final SketchIterator<S> it = sketchIn.iterator();
+
+      final SketchIterator<S> it = nextSketch.iterator();
       while (it.next()) {
-        final S summary = sketch_.find(it.getKey());
-        if (summary != null) {
-          matchKeys[matchCount] = it.getKey();
-          if (matchSummaries == null) {
-            matchSummaries = (S[]) Array.newInstance(summary.getClass(), matchSize);
-          }
-          matchSummaries[matchCount] =
-              summarySetOps_.intersection(summary, it.getSummary());
-          matchCount++;
+        final long key = it.getKey();
+        if (key >= thetaLong_) { continue; }
+        final int index = hashSearch(hashTables_.keys_, hashTables_.lgTableSize_, key);
+        if (index < 0) { continue; }
+        //Copy the intersecting items from local hashTables_
+        // sequentially into local matchKeys_ and matchSummaries_
+        final S mySummary = hashTables_.summaries_[index];
+
+        if (matchSummaries == null) {
+          matchSummaries = (S[]) Array.newInstance(mySummary.getClass(), maxMatchSize);
         }
+        matchKeys[matchCount] = key;
+        matchSummaries[matchCount] = summarySetOps_.intersection(mySummary, it.getSummary());
+        matchCount++;
+      }
+      hashTables_.fromArrays(matchKeys, matchSummaries, matchCount);
+    }
+  }
+
+  /**
+   * Updates the internal set by intersecting it with the given Theta sketch
+   * @param sketchIn input Theta Sketch to intersect with the internal state.
+   * @param summary the given proxy summary for the Theta Sketch, which doesn't have one.
+   */
+  public void update(final org.apache.datasketches.theta.Sketch sketchIn, final S summary) {
```

Comment:
This is a good catch!  Thank you. Will Fix.

Reply:
BTW, I chose to throw an exception here, because I can't figure out what else to do.  I can't assume a default summary since I have no idea what the user would consider to be a default.